### PR TITLE
Add utility class to wrap images as 1D collections

### PR DESCRIPTION
**Pull request by ctrueden**
This is useful to pass images to APIs that work with collections, such as Guava's `com.google.common.math.Quantiles`.

See also [this post on the Image.sc Forum](https://forum.image.sc/t/imglib2-development-a-question-about-img-t/23186/8).

### CODE REVIEW

**Comment by ctrueden** [comment]
A couple of notes.

1. I just noticed @acardona wrote a similar thing on the [`collection` branch](https://github.com/imglib/imglib2/commits/collection). But it doesn't look like a PR was ever filed. Also, that work always maps `IterableInterval<T>` and `RandomAccessibleInterval<T>` to the same type `Collection<T>`, without a way to bake in a conversion function. But if this PR merges, and @acardona is happy with it, then we should probably also delete the `collection` branch, since it would not be necessary anymore.

2. There is a subtle (to me :wink:) bug in the current implementation, relating to images with non-flat iteration order. You can see the problem by adding the following code to the beginning of the `testDoubleList()` method:
    ```java
    final Img< DoubleType > imgDouble = new CellImgFactory<>( new DoubleType() ).create( 30, 40, 50 );
    double v = 0; for ( final DoubleType t : imgDouble ) t.set( v++ );
    ```
    The issue is that this code uses `IntervalIndexer` to map between 1D and nD coordinates, but that function uses flat iteration order. So for things like `CellImg`, the original image iteration order (not flat) does not match the 1D rasterization order of the wrapped collection (flat). Is there an efficient way to introspect a `CellImg` to discern its 1D-to-nD ordering? If not, I am not sure how to solve this problem.

File: src/test/java/net/imglib2/util/FlatCollectionsTest.java [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+
+	private final Img< DoubleType > imgDouble = ArrayImgs.doubles( new double[] { 0.1, 0.02, 0.003, 0.0004, 0.00005, 0.000006, 0.0000007, 0.00000008, 0.000000009, 0, -0.9, -0.08 }, 3, 4 );
+
+	private final Img< IntType > imgInt = ArrayImgs.ints( new int[] { 1, 10, 100, 1_000, 10_000, 100_000, 1_000_000, 10_000_000, 100_000_000, 1_000_000_000, -1_000_000_000, -100_000_000 }, 3, 4 );
+
+	private final Img< LongType > imgLong = ArrayImgs.longs( new long[] { 1, 10, 100, 1_000, 10_000, 100_000, 1_000_000, 10_000_000, 100_000_000, 1_000_000_000, 10_000_000_000L, 100_000_000_000L }, 3, 4 );
+
+	private final Img< ShortType > imgShort = ArrayImgs.shorts( new short[] { 32767, 32766, 32765, 32764, 32763, 10000, 10001, 10002, 10003, -32768, -32767, -32766 }, 2, 3, 2 );
+
+	private final Img< UnsignedLongType > imgBig = ArrayImgs.unsignedLongs( new long[] { 2 * Long.MAX_VALUE, 3 * Long.MAX_VALUE, 4 * Long.MAX_VALUE, Long.MIN_VALUE, Long.MIN_VALUE / 2, Long.MIN_VALUE / 3 }, 3, 2 );
```
[the remark this answers is not in the document]
Comment (ctrueden) [reply src/test/java/net/imglib2/util/FlatCollectionsTest.java]: There is no performance issue with these tests. Testing more edge cases is good. I see no benefit in cutting down these tests.

**Comment by ctrueden** [comment]
Thanks for the review, @maarzt!

I force-pushed an update with the following improvements:

1. I added javadoc to all public methods.
2. I renamed the `ii` parameters to `image`.

I did not (yet) change either of the following:

* Request to remove the byte and short signatures.
* Request to make the testing more minimal.

I think it is a mistake to cut down the testing. Fewer edge cases would be covered. The tests as written are simple to understand and not burdensome to maintain.

Regarding removal of the byte and short signatures: generally speaking, I agree with you that less code is better. Every line of code is a liability, as they say. However, in this case, there are 8 primitive types, and each of them corresponds well to a particular subset of the ImgLib2 type hierarchy, as written here. In my opinion, removing byte and short here would make only make this code more confusing to newcomers, who would see that _most_ but not _all_ of the primitive types have corresponding signatures, and wonder why not. These methods are a convenience, only one line each, so near zero burden to maintain. But replicating them via the general method is potentially error-prone, because each one-line conversion embeds the correct method call (e.g. `t.get()` vs. `.getInteger()` depending on the type) to access the underlying data most directly.

Finally, I disagree that this utility class would be better served in imglib2-algorithm, for the following reasons:

* Other conversion methods such as `Converters` and `Views` exist in imglib2 core, not imglib2-algorithm.
* The `imglib2-algorithm` library brings in many dependencies, none of which are necessary for `FlatCollections`.

File: src/test/java/net/imglib2/util/FlatCollectionsTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (maarzt) [reply src/test/java/net/imglib2/util/FlatCollectionsTest.java]: I also see no performance issue. My points are these:
1. The tests could be simpler. A simple test, makes maintenance easier. When it fails, finding the bug is easier.
2. The test implement the same algorithm as the tested code. If there's a bug with for example the usage of IntervalIndexer, the test want show that, because it would compare the wrong result against an equally wrong expected result. So having a test, which is simple enough to not use for example IntervalIndexer is actually a benefit.
3. There are better places to put test ```getRealDouble``` and similar methods. See PR https://github.com/imglib/imglib2/pull/249 fixes a bug in Unsigned128BitInteger.getRealDouble and also adds tests.

[the remark this answers is not in the document]
Comment (maarzt) [reply src/test/java/net/imglib2/util/FlatCollectionsTest.java]: On the other side. The tests can stay as the are they are good enought

**Comment by maarzt** [comment]
After thinking about it a little bit more. I agree that this PR is mostly fine as it is. Tests are ok, having all the methods is ok, having it as part of imglib2 is fine for me too. But I still have some worries:
* About Byte and Short. I see how supporting byte and short is more inviting to newcomers. At the same time I'm worried, because we only support the rarely used signed `ByteType` and not the much more common `UnsignedByteType`. So supporting `ByteType` but not the `UnsignedByteType` seems to be even more confusing than only having `IntegerType`???
* More javadoc is needed: The javadoc needs to state that these methods have low performance, because of there bad memory footprint related to the boxed primitives. And that the collection don't guaranty any particular iteration order.

**Comment by ctrueden** [comment]
I understand what you are saying about the tests. The way they are written, they are regression tests. They aim to be a little more thorough (exercising multiple methods of `Collection` and `List`), at the expense of layers of indirection with the helper methods and reuse of images. If you want to rewrite them to your liking, go for it. I won't be offended. But I don't have time personally.

I pushed two more commits: one removing the byte and short wrappers, and one adding a note to the javadoc about the iteration order.

Regarding performance of the wrapped collections, there should be virtually no memory overhead. Yes, there is boxing. However, it's not like `ArrayList<Double>` where there will be `size()` `Double` objects in existence simultaneously. If you really want me to benchmark it I will, but I do not expect the memory use to be too bad. Only time performance potentially impacted compared to primitives, as transiently creating all those boxed objects takes CPU time too.

**Comment by ctrueden** [comment]
Thanks @maarzt!
